File: src/main/java/xyz/chalky/taboo/events/GuildEvents.java
package xyz.chalky.taboo.events;

import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.events.channel.ChannelCreateEvent;
import net.dv8tion.jda.api.events.channel.ChannelDeleteEvent;
import net.dv8tion.jda.api.events.channel.update.GenericChannelUpdateEvent;
import net.dv8tion.jda.api.events.role.RoleCreateEvent;
import net.dv8tion.jda.api.events.role.RoleDeleteEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import xyz.chalky.taboo.database.model.Config;
import xyz.chalky.taboo.database.repository.ConfigRepository;

import java.awt.*;
import java.time.Instant;

@Component
public class GuildEvents extends ListenerAdapter {

    @Autowired private ConfigRepository configRepository;

    // Role Events

    @Override
    public void onRoleCreate(@NotNull RoleCreateEvent event) {
        Role role = event.getRole();
        Guild guild = event.getGuild();
        Long logId = configRepository.findById(guild.getIdLong()).map(Config::getLogChannelId).orElse(null);
        if (logId == null) return;
        TextChannel log = guild.getTextChannelById(logId);
        if (log == null) return;
        MessageEmbed embed = new EmbedBuilder()
                .setTitle("Role Created")
                .setDescription(String.format("Role: %s", role.getAsMention()))
                .setColor(role.getColorRaw())
                .setFooter(String.format("Role ID: %s", role.getId()))
                .setTimestamp(Instant.now())
                .build();
        log.sendMessageEmbeds(embed).queue();
    }

    @Override
    public void onRoleDelete(@NotNull RoleDeleteEvent event) {
        Role role = event.getRole();
        Guild guild = event.getGuild();
        Long logId = configRepository.findById(guild.getIdLong()).map(Config::getLogChannelId).orElse(null);
        if (logId == null) return;
        TextChannel log = guild.getTextChannelById(logId);
        if (log == null) return;
        MessageEmbed embed = new EmbedBuilder()
                .setTitle("Role Deleted")
                .setDescription(String.format("Role: %s", role.getAsMention()))
                .setColor(role.getColorRaw())
                .setFooter(String.format("Role ID: %s", role.getId()))
                .setTimestamp(Instant.now())
                .build();
        log.sendMessageEmbeds(embed).queue();
    }

    // Channel Events


    @Override
    public void onChannelCreate(@NotNull ChannelCreateEvent event) {
        Channel channel = event.getChannel();
        Guild guild = event.getGuild();
        Long logId = configRepository.findById(guild.getIdLong()).map(Config::getLogChannelId).orElse(null);
        if (logId == null) return;
        TextChannel log = guild.getTextChannelById(logId);
        if (log == null) return;
        MessageEmbed embed = new EmbedBuilder()
                .setTitle("Channel Created")
                .setDescription(String.format("Channel: %s", channel.getAsMention()))
                .setColor(Color.CYAN)
                .setFooter(String.format("Channel ID: %s", channel.getId()))
                .setTimestamp(Instant.now())
                .build();
        log.sendMessageEmbeds(embed).queue();
    }

    @Override
    public void onChannelDelete(@NotNull ChannelDeleteEvent event) {
        Channel channel = event.getChannel();
        Guild guild = event.getGuild();
        Long logId = configRepository.findById(guild.getIdLong()).map(Config::getLogChannelId).orElse(null);
        if (logId == null) return;
        TextChannel log = guild.getTextChannelById(logId);
        if (log == null) return;
        MessageEmbed embed = new EmbedBuilder()
                .setTitle("Channel Deleted")
                .setDescription(String.format("Channel: %s", channel.getAsMention()))
                .setColor(Color.CYAN)
                .setFooter(String.format("Channel ID: %s", channel.getId()))
                .setTimestamp(Instant.now())
                .build();
        log.sendMessageEmbeds(embed).queue();
    }

    @Override
    public void onGenericChannelUpdate(@NotNull GenericChannelUpdateEvent<?> event) {
        Channel channel = event.getChannel();
        Guild guild = event.getGuild();
        Long logId = configRepository.findById(guild.getIdLong()).map(Config::getLogChannelId).orElse(null);
        if (logId == null) return;
        TextChannel log = guild.getTextChannelById(logId);
        if (log == null) return;
        MessageEmbed embed = new EmbedBuilder()
                .setTitle("Channel Updated")
                .setDescription(String.format("Channel: %s", channel.getAsMention()))
                .setColor(Color.CYAN)
                .setFooter(String.format("Channel ID: %s", channel.getId()))
                .setTimestamp(Instant.now())
                .build();
        log.sendMessageEmbeds(embed).queue();
    }

}

File: src/main/kotlin/xyz/chalky/taboo/events/GuildEvents.kt
package xyz.chalky.taboo.events

import dev.minn.jda.ktx.Embed
import net.dv8tion.jda.api.events.channel.ChannelCreateEvent
import net.dv8tion.jda.api.events.channel.ChannelDeleteEvent
import net.dv8tion.jda.api.events.channel.update.GenericChannelUpdateEvent
import net.dv8tion.jda.api.events.role.RoleCreateEvent
import net.dv8tion.jda.api.events.role.RoleDeleteEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter
import xyz.chalky.taboo.util.getActionLogId
import java.awt.Color
import java.time.Instant

class GuildEvents : ListenerAdapter() {

    // Role Events

    override fun onRoleCreate(event: RoleCreateEvent) {
        val role = event.role
        val guild = event.guild
        val actionLogId = getActionLogId(guild) ?: return
        val actionLog = guild.getTextChannelById(actionLogId) ?: return
        actionLog.sendMessageEmbeds(
            Embed {
                title = "Role Created"
                description = "Role: ${role.asMention}"
                color = role.color.hashCode()
                footer {
                    name = "Role ID: ${role.id}"
                }
                timestamp = Instant.now()
            }
        ).queue()
    }

    override fun onRoleDelete(event: RoleDeleteEvent) {
        val role = event.role
        val guild = event.guild
        val actionLogId = getActionLogId(guild) ?: return
        val actionLog = guild.getTextChannelById(actionLogId) ?: return
        actionLog.sendMessageEmbeds(
            Embed {
                title = "Role Deleted"
                description = "Role: ${role.asMention}"
                color = role.color.hashCode()
                footer {
                    name = "Role ID: ${role.id}"
                }
                timestamp = Instant.now()
            }
        ).queue()
    }

    // Channel Events

    override fun onChannelCreate(event: ChannelCreateEvent) {
        val channel = event.channel
        val guild = event.guild
        val actionLogId = getActionLogId(guild) ?: return
        val actionLog = guild.getTextChannelById(actionLogId) ?: return
        actionLog.sendMessageEmbeds(
            Embed {
                title = "Channel Created"
                description = "Channel: ${channel.asMention}"
                color = Color.CYAN.hashCode()
                footer {
                    name = "Channel ID: ${channel.id}"
                }
                timestamp = Instant.now()
            }
        ).queue()
    }

    override fun onChannelDelete(event: ChannelDeleteEvent) {
        val channel = event.channel
        val guild = event.guild
        val actionLogId = getActionLogId(guild) ?: return
        val actionLog = guild.getTextChannelById(actionLogId) ?: return
        actionLog.sendMessageEmbeds(
            Embed {
                title = "Channel Deleted"
                description = "Channel: ${channel.asMention}"
                color = Color.CYAN.hashCode()
                footer {
                    name = "Channel ID: ${channel.id}"
                }
                timestamp = Instant.now()
            }
        ).queue()
    }

    override fun onGenericChannelUpdate(event: GenericChannelUpdateEvent<*>) {
        val channel = event.channel
        val guild = event.guild
        val actionLogId = getActionLogId(guild) ?: return
        val actionLog = guild.getTextChannelById(actionLogId) ?: return
        actionLog.sendMessageEmbeds(
            Embed {
                title = "Channel Updated"
                description = "Channel: ${channel.asMention}"
                color = Color.CYAN.hashCode()
                footer {
                    name = "Channel ID: ${channel.id}"
                }
                timestamp = Instant.now()
            }
        ).queue()
    }

}